include/deal.II/base/memory_space_data.h

```diff
@@ -25,12 +25,124 @@
 #include <functional>
 #include <memory>
 
+#ifdef DEAL_II_USE_KOKKOS_BACKEND
+#  include <Kokkos_Core.hpp>
+#endif
+
 DEAL_II_NAMESPACE_OPEN
 
 /**
  */
 namespace MemorySpace
 {
+#ifdef DEAL_II_USE_KOKKOS_BACKEND
+  /**
+   * Data structure
+   */
+  template <typename Number, typename MemorySpace>
+  struct MemorySpaceData
+  {
+    MemorySpaceData()
+      : values("memoryspace data", 0)
+    {}
+
+    bool
+    has_data()
+    {
+      return values.extent(0) > 0;
+    }
+
+    Number *
+    data()
+    {
+      return values.data();
+    }
+
+    const Number *
+    data() const
+    {
+      return values.data();
+    }
+
+    Number &
+    operator()(unsigned int i)
+    {
+      return values(i);
+    }
+
+    const Number &
+    operator()(unsigned int i) const
+    {
+      return values(i);
+    }
+
+    /**
+     * Copy the active data (values for Host and values_dev for CUDA) to @p begin.
+     * If the data is on the device it is moved to the host.
+     */
+    void
+    copy_to(Number *begin, std::size_t n_elements)
+    {
+      Assert(n_elements <= values.extent(0), ExcMessage("aa"));
+      Kokkos::parallel_for(
+        "MemorySpaceData::copy_to",
+        Kokkos::RangePolicy<typename MemorySpace::execution_space>(0,
+                                                                   n_elements),
+        KOKKOS_LAMBDA(const int i) { *(begin + i) = values(i); });
+    }
+
+    /**
+     * Copy the data in @p begin to the active data of the structure (values for
+     * Host and values_dev for CUDA). The pointer @p begin must be on the host.
+     */
+    void
+    copy_from(Number *begin, std::size_t n_elements)
+    {
+      Assert(n_elements <= values.extent(0), ExcMessage("bb"));
+      Kokkos::parallel_for(
+        "MemorySpaceData::copy_from",
+        Kokkos::RangePolicy<typename MemorySpace::execution_space>(0,
+                                                                   n_elements),
+        KOKKOS_LAMBDA(const int i) { values(i) = *(begin + i); });
+    }
+
+    /**
+     * Kokkos View to the data
+     */
+    Kokkos::View<Number *, MemorySpace> values;
+
+    /**
+     * Pointers to the data of the processes sharing the same memory.
+     */
+    std::vector<ArrayView<const Number>> values_sm;
+
+    /**
+     * Pointer to data on the host. The pointer points to the same data than
+     * values when using shared memory. Otherwise it is not set.
+     */
+    // The pointer is shared so that MemorySpaceData can be copied and
+    // MemorySpaceData::values can be used in Kokkos::parallel_for.
+    std::shared_ptr<Number[]> values_sm_ptr;
```

Comment:
Can you comment on who owns the data? This object, or the `values` object?

Reply:
This is the messy part. When you don't use shared MPI the data is owned by `values`. When you use shared MPI, the Kokkos::View needs to be changed to a non-owning View because it cannot use `MPI_Win_Allocate` and `MPI_Win_Free`. When you don't use Kokkos, we allocate the memory, give the pointer to `value` and then we let `value` free the memory. Because we cannot do that anymore, I need another object tho free memory that's what `values_sm_ptr` does. To answer your question if you don't use shared MPI, `values` owns the data. If you use shared MPI, `values_sm_ptr` owns the data.

Reply:
Yes, the idea is that the class owns the data and the destructor will free the memory. From the user's perspective, there is no difference if you use shared MPI or not.